src/Solvers/multigrid_solver.jl

```diff
@@ -0,0 +1,163 @@
+                    )
+
+Returns a MultigridSolver that solves the linear equation
+``A x = b`` using a multigrid method, where `A * x` is
+determined by `linear_operation!`
+
+`linear_operation!` is a function with signature `linear_operation!(Ax, x, args...)` 
+that calculates `A * x` for given `x` and stores the result in `Ax`.
+
+
+The solver is used by calling
+
+```
+solve!(x, solver::MultigridSolver, b; kwargs...)
+```
+
+for `solver`, right-hand side `b`, solution `x`, and optional keyword arguments `kwargs...`.
+
+Arguments
+=========
+
+* `template_field`: Dummy field that is the same type and size as `x` and `b`, which
+                    is used to infer the `architecture`, `grid`, and to create work arrays
+                    that are used internally by the solver.
+
+* `maxiter`: Maximum number of iterations the solver may perform before exiting.
+
+* `reltol, abstol`: Relative and absolute tolerance for convergence of the algorithm.
+                    The iteration stops when `norm(A * x - b) < tolerance`.
+
+* `amg_algorithm`: Algebraic Multigrid algorithm defining mapping between different grid spacings
+
+!!! compat "Multigrid solver on GPUs"
+    Currently Multigrid solver is only supported on CPUs.
+"""
+function MultigridSolver(linear_operation!::Function,
+                         args...;
+                         template_field::AbstractField,
+                         maxiter = prod(size(template_field)),
+                         reltol = sqrt(eps(eltype(template_field.grid))),
+                         abstol = 0,
+                         amg_algorithm = RugeStubenAMG(),
+                         )
+
+    arch = architecture(template_field)
+
+    arch = GPU() && error("Multigrid solver is only supported on CPUs.")
+
+    matrix = create_matrix(template_field, linear_operation!, args...)
+
+    Nx, Ny, Nz = size(template_field)
+
+    FT = eltype(template_field.grid)
+
+    b_array = arch_array(arch, zeros(FT, Nx * Ny * Nz))
+    x_array = arch_array(arch, zeros(FT, Nx * Ny * Nz))
+
+    return MultigridSolver(arch,
+                           template_field.grid,
+                           matrix,
+                           FT(abstol),
+                           FT(reltol),
+                           maxiter,
+                           amg_algorithm,
+                           x_array,
+                           b_array
+                           )
+end
+
+
+# For free surface without Δt
+function initialize_matrix(template_field, ::Function, ::Any , ::Any, ::Any, ::Nothing)
+    Nx, Ny, Nz = size(template_field)
+    return spzeros(eltype(template_field.grid), Nx*Ny*Nz, Nx*Ny*Nz)
+end
+
+function initialize_matrix(template_field, linear_operator!, args...)
+    Nx, Ny, Nz = size(template_field)
+    A = spzeros(eltype(template_field.grid), Nx*Ny*Nz, Nx*Ny*Nz)
+
+    create_matrix!(A, template_field, linear_operator!, args...)
+    
+    return A
+end
+
```

Comment:
```suggestion
function initialize_matrix(template_field)
    Nx, Ny, Nz = size(template_field)
    return spzeros(eltype(template_field.grid), Nx*Ny*Nz, Nx*Ny*Nz)
end

```

Reply:
see where I'm heading with this? 
initialize just allocated the matrix of appropriate size...
then `fill_matrix!` fills it up...
this way we don't need two methods for when `Δt = nothing`... right?

Reply:
I think the issue with this is the type signature was allowing us to determine whether or not we should initialise the elements (ie is Δt initialised)

Reply:
well... but the function name will do that

For the free-surface solver, the constructor will only call `initialize_matrix` and that's done, right? The filling is done within `solve!`. Right?

Reply:
Despite our previous conversation I think I just realised why doing it like this is messy
- `initialize_matrix` and `fill_matrix_elements!` are both called from the `MultigridSolver` constructor (rather than `solve!` as we only want to initialize the matrix once if possible)
- The `MGImplicitFreeSurfaceSolver` constructor calls the `MultigridSolver` constructor but does not want to call `fill_matrix_elements!` 

Therefore we need some way to differenciate in the `MultigridSolver` constructor whether we are in the implicit free surface case or not - the current way of doing so is to use the type signature, there are clearly other ways but are they more or less messy than creating another function?

Reply:
OK then.

Let's just put the `create_matrix` method that is specific for the FreeSurface solver in the file with the free surface solver.

Reply:
I don't know how to import the free surface specific `initialize_matrix` from `mg_implicit_free_surface_solver.jl` to `multigrid_solver.jl`, I tried 
```using Oceananigans.Models.HydrostaticFreeSurfaceModels: initialize_matrix```
but I get
```LoadError: UndefVarError: Models not defined```

Reply:
Why does the `Solvers/multigrid_solver.jl` need to know anything about the free surface?
I thought it'd be the other way around. Like define the function in `Solvers/multigrid_solver.jl` and then inside `mg_implicit_free_surface_solver.jl` import it and add a method. No?

Reply:
Perhaps just leave it where it is :) I just added a comment suggestion that would make it clear what's it for.